File: archive/polymath-web-lib/traversal.ts
/**
 * Traversal engine for domain recommendations
 * Ported from pm/core/traversal.py
 */

import type {
  DomainWithProgress,
  TraversalRecommendation,
  FunctionSlot,
  TraversalPhase,
} from '@/types';
import { HUB_DOMAIN_IDS } from '@/types';
import { getBranchDistance } from './distance';
import {
  SUNDAY,
  HUB_TARGET_BOOKS,
  MIN_BISOCIATION_DISTANCE,
  TOP_CANDIDATES_POOL,
  DISTANT_INTERLEAVE_BONUS,
  DEFAULT_STRENGTH_BRANCH,
  getBranchId,
} from './constants';

/**
 * Get the next function slot for a domain based on books read
 */
export function getNextSlot(booksRead: number): FunctionSlot {
  if (booksRead === 0) return 'FND';
  if (booksRead === 1) return 'HRS';
  if (booksRead === 2) return 'ORT';
  if (booksRead === 3) return 'FRN';
  if (booksRead === 4) return 'HST';
  return 'BRG';
}

/**
 * Get slot display name
 */
export function getSlotName(slot: FunctionSlot): string {
  const names: Record<FunctionSlot, string> = {
    FND: 'Foundation',
    HRS: 'Heresy',
    ORT: 'Orthodoxy',
    FRN: 'Frontier',
    HST: 'History',
    BRG: 'Bridge',
  };
  return names[slot];
}

/**
 * Main traversal engine class
 */
export class TraversalEngine {
  private hubTargetBooks: number;
  private expertDomains: string[];
  private minBisociationDistance: number;

  constructor(config: {
    hubTargetBooks?: number;
    expertDomains?: string[];
    minBisociationDistance?: number;
  } = {}) {
    this.hubTargetBooks = config.hubTargetBooks ?? HUB_TARGET_BOOKS;
    this.expertDomains = config.expertDomains ?? [];
    this.minBisociationDistance = config.minBisociationDistance ?? MIN_BISOCIATION_DISTANCE;
  }

  /**
   * Get strength branches from expert domains
   */
  private getStrengthBranches(domains: DomainWithProgress[]): Set<string> {
    const branches = new Set(
      [...this.expertDomains, ...domains.filter((d) => d.is_expert).map((d) => d.domain_id)]
        .map(getBranchId)
    );

    if (branches.size === 0) {
      branches.add(DEFAULT_STRENGTH_BRANCH);
    }

    return branches;
  }

  /**
   * Get next reading recommendation
   */
  recommendNext(
    domains: DomainWithProgress[],
    recentDomainIds: string[] = [],
    phase: TraversalPhase = 'hub-completion'
  ): TraversalRecommendation | null {
    const dayOfWeek = new Date().getDay();
    const isInterleaveDay = dayOfWeek === SUNDAY;

    if (phase === 'hub-completion') {
      return this.recommendHubCompletion(domains, recentDomainIds, isInterleaveDay);
    } else if (phase === 'bisociation') {
      return this.recommendBisociation(domains, recentDomainIds, isInterleaveDay);
    } else {
      return this.recommendProblemDriven(domains, recentDomainIds);
    }
  }

  /**
   * Hub completion phase recommendation
   */
  private recommendHubCompletion(
    domains: DomainWithProgress[],
    recentDomainIds: string[],
    isInterleaveDay: boolean
  ): TraversalRecommendation | null {
    // Check if it's a distant interleave day
    if (isInterleaveDay) {
      const distant = this.findDistantDomain(domains, recentDomainIds);
      if (distant) return distant;
    }

    // Find incomplete hub domains
    const hubDomains = domains.filter(
      (d) =>
        HUB_DOMAIN_IDS.includes(d.domain_id as typeof HUB_DOMAIN_IDS[number]) &&
        d.books_read < this.hubTargetBooks &&
        !recentDomainIds.includes(d.domain_id)
    );

    if (hubDomains.length === 0) {
      // All hubs complete, fall back to distant domain
      return this.findDistantDomain(domains, recentDomainIds);
    }

    // Prioritize closest to completion
    hubDomains.sort((a, b) => b.books_read - a.books_read);
    const domain = hubDomains[0];

    return {
      domain,
      slot: getNextSlot(domain.books_read),
      reason: `Start hub domain with ${getNextSlot(domain.books_read)} slot (${this.hubTargetBooks - domain.books_read} books to complete)`,
      phase: 'hub-completion',
    };
  }

  /**
   * Bisociation phase recommendation
   */
  private recommendBisociation(
    domains: DomainWithProgress[],
    recentDomainIds: string[],
    isInterleaveDay: boolean
  ): TraversalRecommendation | null {
    // On interleave days, recommend strength domain
    if (!isInterleaveDay) {
      return this.findStrengthDomain(domains, recentDomainIds);
    }

    return this.findDistantDomain(domains, recentDomainIds);
  }

  /**
   * Problem-driven phase recommendation
   */
  private recommendProblemDriven(
    domains: DomainWithProgress[],
    recentDomainIds: string[]
  ): TraversalRecommendation | null {
    // For now, just recommend an untouched domain
    const untouched = domains.filter(
      (d) =>
        d.status === 'untouched' &&
        !recentDomainIds.includes(d.domain_id)
    );

    if (untouched.length === 0) return null;

    // Random selection
    const domain = untouched[Math.floor(Math.random() * untouched.length)];

    return {
      domain,
      slot: 'FND',
      reason: 'Explore new domain for problem-driven learning',
      phase: 'problem-driven',
    };
  }

  /**
   * Find a domain at maximum distance from user's strengths
   */
  private findDistantDomain(
    domains: DomainWithProgress[],
    recentDomainIds: string[]
  ): TraversalRecommendation | null {
    const strengthBranches = this.getStrengthBranches(domains);

    // Score domains by distance from strengths
    const candidates = domains
      .filter((d) => !recentDomainIds.includes(d.domain_id))
      .map((d) => {
        const branchId = getBranchId(d.domain_id);
        const minDistance = Math.min(
          ...Array.from(strengthBranches).map((sb) =>
            getBranchDistance(branchId, sb)
          )
        );

        // Bonus for untouched
        const bonus = d.status === 'untouched' ? DISTANT_INTERLEAVE_BONUS : 0;

        return { domain: d, score: minDistance + bonus };
      })
      .filter((c) => c.score >= this.minBisociationDistance);

    if (candidates.length === 0) return null;

    // Sort by score descending, take from top candidates randomly
    candidates.sort((a, b) => b.score - a.score);
    const topCandidates = candidates.slice(0, TOP_CANDIDATES_POOL);

    // Safety check: ensure we have candidates after slicing
    if (topCandidates.length === 0) return null;

    const selected = topCandidates[Math.floor(Math.random() * topCandidates.length)];

    // Safety check: ensure selection succeeded
    if (!selected || !selected.domain) return null;

    return {
      domain: selected.domain,
      slot: getNextSlot(selected.domain.books_read),
      reason: `Distant domain interleave (distance: ${Math.floor(selected.score)})`,
      phase: 'hub-completion',
    };
  }

  /**
   * Find a domain in user's strength areas
   */
  private findStrengthDomain(
    domains: DomainWithProgress[],
    recentDomainIds: string[]
  ): TraversalRecommendation | null {
    const strengthBranches = this.getStrengthBranches(domains);

    // Find domains in strength branches that need more reading
    const candidates = domains.filter(
      (d) =>
        strengthBranches.has(getBranchId(d.domain_id)) &&
        d.status !== 'expert' &&
        !recentDomainIds.includes(d.domain_id)
    );

    if (candidates.length === 0) return null;

    // Prioritize by books read (deepen existing)
    candidates.sort((a, b) => b.books_read - a.books_read);
    const domain = candidates[0];

    return {
      domain,
      slot: getNextSlot(domain.books_read),
      reason: 'Deepen strength domain during bisociation phase',
      phase: 'bisociation',
    };
  }

  /**
   * Check if hub completion phase is done
   */
  isHubCompletionDone(domains: DomainWithProgress[]): boolean {
    const hubs = domains.filter((d) =>
      HUB_DOMAIN_IDS.includes(d.domain_id as typeof HUB_DOMAIN_IDS[number])
    );

    // Empty array .every() returns true - must verify we found all hub domains
    if (hubs.length === 0 || hubs.length < HUB_DOMAIN_IDS.length) {
      return false; // Missing hub domains in data
    }

    return hubs.every((d) => d.books_read >= this.hubTargetBooks);
  }
}

// Default traversal engine instance
export const traversalEngine = new TraversalEngine();

File: archive/pm/core/traversal.py
"""Traversal engine for Polymath Engine.

Implements the recommendation logic for what to read next.
"""

from dataclasses import dataclass
from datetime import date, timedelta
from enum import Enum
from typing import Optional

from pm.config import TraversalConfig
from pm.core.domain import Domain, DomainStatus, FunctionSlot
from pm.data.distances import get_branch_distance


class TraversalPhase(Enum):
    """Current phase of the traversal strategy."""

    HUB_COMPLETION = "hub-completion"
    PROBLEM_DRIVEN = "problem-driven"
    BISOCIATION = "bisociation"


@dataclass
class TraversalRecommendation:
    """A reading recommendation from the traversal engine."""

    domain: Domain
    slot: FunctionSlot
    reason: str
    phase: TraversalPhase
    is_distant_interleave: bool = False
    distance_from_strength: int = 0
    priority: int = 0  # Lower is higher priority


class TraversalEngine:
    """Engine for generating reading recommendations."""

    def __init__(self, config: TraversalConfig):
        """Initialize traversal engine.

        Args:
            config: Traversal configuration.
        """
        self.config = config
        self.current_phase = TraversalPhase.HUB_COMPLETION
        # Normalize config field names
        self.books_per_hub = getattr(config, 'hub_target_books', 4)
        self.cooldown_days = getattr(config, 'max_domain_repeat_window', 14)
        self.min_distant_distance = getattr(config, 'bisociation_min_distance', 3)
        self.distant_interleave_day = 6  # Sunday by default

    def recommend_next(
        self,
        domains: list[Domain],
        recent_domain_ids: list[str],
        week_day: int = 0,
    ) -> Optional[TraversalRecommendation]:
        """Get the next reading recommendation.

        Args:
            domains: All domains from the vault.
            recent_domain_ids: Domain IDs read in last N days.
            week_day: Day of the week (0=Monday, 6=Sunday).

        Returns:
            TraversalRecommendation or None if no recommendation.
        """
        if self.current_phase == TraversalPhase.HUB_COMPLETION:
            return self._recommend_hub_completion(domains, recent_domain_ids, week_day)
        elif self.current_phase == TraversalPhase.PROBLEM_DRIVEN:
            return self._recommend_problem_driven(domains, recent_domain_ids)
        else:
            return self._recommend_bisociation(domains, recent_domain_ids, week_day)

    def _recommend_hub_completion(
        self,
        domains: list[Domain],
        recent_domain_ids: list[str],
        week_day: int,
    ) -> Optional[TraversalRecommendation]:
        """Recommend based on hub completion strategy.

        Logic:
        1. Prioritize incomplete hub domains
        2. Weekly interleave with distant domain (distance >= 3)
        3. Respect cooldown period for recently read domains
        """
        # Check if today should be a distant interleave day
        # Default: every 7th day is distant interleave
        is_distant_day = week_day == self.distant_interleave_day

        if is_distant_day:
            rec = self._find_distant_domain(domains, recent_domain_ids)
            if rec:
                return rec

        # Otherwise, find next hub to work on
        hub_domains = [d for d in domains if d.is_hub]

        # Filter out recently read (cooldown)
        eligible_hubs = [
            d for d in hub_domains
            if d.domain_id not in recent_domain_ids
            and d.books_read < self.books_per_hub
        ]

        if not eligible_hubs:
            # All hubs complete or on cooldown, try distant domain
            return self._find_distant_domain(domains, recent_domain_ids)

        # Prioritize by: closest to completion, then by domain_id for consistency
        eligible_hubs.sort(key=lambda d: (-d.books_read, d.domain_id))

        best_hub = eligible_hubs[0]
        slot = best_hub.next_slot()

        return TraversalRecommendation(
            domain=best_hub,
            slot=slot,
            reason=self._hub_reason(best_hub),
            phase=TraversalPhase.HUB_COMPLETION,
            is_distant_interleave=False,
            priority=0,
        )

    def _recommend_problem_driven(
        self,
        domains: list[Domain],
        recent_domain_ids: list[str],
    ) -> Optional[TraversalRecommendation]:
        """Recommend based on problem-driven strategy.

        This requires active problems to be defined. Falls back to hub completion.
        """
        # TODO: Implement problem-driven logic when problems are loaded
        # For now, fall back to hub completion
        return self._recommend_hub_completion(domains, recent_domain_ids, 0)

    def _recommend_bisociation(
        self,
        domains: list[Domain],
        recent_domain_ids: list[str],
        week_day: int,
    ) -> Optional[TraversalRecommendation]:
        """Recommend based on bisociation strategy.

        Weekly rhythm:
        - Days 0-2: Strength domains (expert or high books_read)
        - Days 3-5: Distant domains (max distance from strength)
        - Day 6: Synthesis day (no new reading)
        """
        if week_day == 6:
            # Synthesis day
            return None

        if week_day < 3:
            # Strength days - work on expert domains or domains with progress
            return self._find_strength_domain(domains, recent_domain_ids)
        else:
            # Distant days
            return self._find_distant_domain(domains, recent_domain_ids)

    def _find_distant_domain(
        self,
        domains: list[Domain],
        recent_domain_ids: list[str],
    ) -> Optional[TraversalRecommendation]:
        """Find a domain at maximum distance from user's strength areas."""
        # Get expert/strength branches
        strength_branches = set()
        for d in domains:
            if d.is_expert or d.books_read >= 2:
                strength_branches.add(d.branch_id)

        if not strength_branches:
            # No strength areas yet, use Engineering (07) as default
            strength_branches = {"07"}

        # Find domains at max distance from all strength branches
        candidates = []
        for d in domains:
            if d.domain_id in recent_domain_ids:
                continue

            # Calculate minimum distance from any strength branch
            min_distance = min(
                get_branch_distance(d.branch_id, sb)
                for sb in strength_branches
            )

            if min_distance >= self.min_distant_distance:
                # Bonus for untouched domains
                bonus = 1 if d.status == DomainStatus.UNTOUCHED else 0
                candidates.append((d, min_distance, bonus))

        if not candidates:
            return None

        # Sort by distance desc, then bonus, then domain_id
        candidates.sort(key=lambda x: (-x[1], -x[2], x[0].domain_id))

        # Pick from top candidates (add some variety)
        best = candidates[0]
        domain = best[0]

        return TraversalRecommendation(
            domain=domain,
            slot=domain.next_slot(),
            reason=f"Distant exploration (distance {best[1]} from your strengths)",
            phase=self.current_phase,
            is_distant_interleave=True,
            distance_from_strength=best[1],
            priority=10,
        )

    def _find_strength_domain(
        self,
        domains: list[Domain],
        recent_domain_ids: list[str],
    ) -> Optional[TraversalRecommendation]:
        """Find a domain in user's strength areas to deepen."""
        # Prefer expert domains or high-progress domains
        candidates = []
        for d in domains:
            if d.domain_id in recent_domain_ids:
                continue

            if d.is_expert or d.books_read >= 2:
                candidates.append(d)

        if not candidates:
            # No strength domains yet, use hubs
            candidates = [d for d in domains if d.is_hub and d.domain_id not in recent_domain_ids]

        if not candidates:
            return None

        # Sort by books_read desc to continue deepening
        candidates.sort(key=lambda d: (-d.books_read, d.domain_id))
        best = candidates[0]

        return TraversalRecommendation(
            domain=best,
            slot=best.next_slot(),
            reason="Deepening strength area" if best.is_expert else "Continuing hub completion",
            phase=self.current_phase,
            is_distant_interleave=False,
            priority=5,
        )

    def _hub_reason(self, domain: Domain) -> str:
        """Generate reason string for hub recommendation."""
        remaining = self.books_per_hub - domain.books_read
        slot = domain.next_slot()

        if domain.books_read == 0:
            return f"Start hub domain with {slot} slot ({remaining} books to complete)"
        else:
            return f"Continue hub ({domain.books_read}/{self.books_per_hub}), next: {slot} slot"

    def set_phase(self, phase: TraversalPhase) -> None:
        """Set the current traversal phase.

        Args:
            phase: New phase to use.
        """
        self.current_phase = phase

    def check_hub_completion(self, domains: list[Domain]) -> bool:
        """Check if all hubs are complete.

        Args:
            domains: All domains.

        Returns:
            True if all hub domains have >= books_per_hub books read.
        """
        hub_domains = [d for d in domains if d.is_hub]
        return all(d.books_read >= self.books_per_hub for d in hub_domains)
